src/server/ua_server.c

```diff
@@ -544,65 +678,174 @@ UA_Server_removeCallback(UA_Server *server, UA_UInt64 callbackId) {
 }
 
 static void
-notifySecureChannelsStopped(UA_Server *server, struct UA_ServerComponent *sc,
-                            UA_LifecycleState state) {
-    if(sc->state == UA_LIFECYCLESTATE_STOPPED &&
-       server->state == UA_LIFECYCLESTATE_STARTED) {
-        sc->notifyState = NULL; /* remove the callback */
-        sc->start(server, sc);
+sercureChannel_delayedClose(void *application, void *context) {
+    UA_Server *server = (UA_Server*)context;
+    UA_GDSTransaction *transaction = server->transaction;
+    UA_Boolean closeRsaMin = false;
+    UA_Boolean closeRsaSha256 = false;
+
+    /* Check which secureChannels are affected and close them */
+    UA_NodeId rsaShaCertificateType = UA_NODEID_NUMERIC(0, UA_NS0ID_RSASHA256APPLICATIONCERTIFICATETYPE);
+    UA_NodeId rsaMinCertificateType = UA_NODEID_NUMERIC(0, UA_NS0ID_RSAMINAPPLICATIONCERTIFICATETYPE);
+
+    for(size_t i = 0; i < transaction->certificateInfosSize; i++) {
+        UA_GDSCertificateInfo certInfo = transaction->certificateInfos[i];
+        if(UA_NodeId_equal(&certInfo.certificateType, &rsaMinCertificateType))
+            closeRsaMin = true;
+        if(UA_NodeId_equal(&certInfo.certificateType, &rsaShaCertificateType))
+            closeRsaSha256 = true;
     }
+
+    UA_NodeId defaultApplicationGroup =
+            UA_NODEID_NUMERIC(0, UA_NS0ID_SERVERCONFIGURATION_CERTIFICATEGROUPS_DEFAULTAPPLICATIONGROUP_TRUSTLIST);
+
+    for(size_t i = 0; i < transaction->certGroupSize; i++) {
+        UA_CertificateGroup certGroup = transaction->certGroups[i];
+        if(!UA_NodeId_equal(&certGroup.certificateGroupId, &defaultApplicationGroup))
+            continue;
+        closeRsaMin = true;
+        closeRsaSha256 = true;
```

Comment:
Why do you hardcode this here? The type can be directly taken from the CertificateInfo in the transaction object.